runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/ExternalTest.java

```diff
@@ -0,0 +1,187 @@
+import org.apache.beam.vendor.guava.v20_0.com.google.common.collect.ImmutableMap;
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+/** Test External transforms. */
+@RunWith(JUnit4.class)
+public class ExternalTest implements Serializable {
+  @Rule public transient TestPipeline testPipeline = TestPipeline.create();
+
+  private static Server expansionServer;
+  private static final int EXPANSION_PORT = 8096;
+  private static final String EXPANSION_ADDR = String.format("localhost:%s", EXPANSION_PORT);
+  private static String pythonServerCommand;
+
+  @BeforeClass
+  public static void setUp() throws IOException {
+    pythonServerCommand = System.getProperty("pythonTestExpansionCommand");
+
+    expansionServer =
+        ServerBuilder.forPort(EXPANSION_PORT).addService(new ExpansionService()).build();
+    expansionServer.start();
+  }
+
+  @AfterClass
+  public static void tearDown() {
+    expansionServer.shutdownNow();
+  }
+
+  @Test
+  @Category({ValidatesRunner.class, UsesCrossLanguageTransforms.class})
+  public void expandSingleTest() {
+    PCollection<Integer> col =
+        testPipeline
+            .apply(Create.of(1, 2, 3))
+            .apply(External.of("simple", new byte[] {}, EXPANSION_ADDR));
+    PAssert.that(col).containsInAnyOrder(2, 3, 4);
+    testPipeline.run();
+  }
+
+  @Test
+  @Category({ValidatesRunner.class, UsesCrossLanguageTransforms.class})
+  public void expandMultipleTest() {
+    PCollection<Integer> pcol =
+        testPipeline
+            .apply(Create.of(1, 2, 3))
+            .apply("add one", External.of("simple", new byte[] {}, EXPANSION_ADDR))
+            .apply(
+                "filter <=3",
+                External.of("le", "3".getBytes(StandardCharsets.UTF_8), EXPANSION_ADDR));
+
+    PAssert.that(pcol).containsInAnyOrder(2, 3);
+    testPipeline.run();
+  }
+
+  @Test
+  @Category({ValidatesRunner.class, UsesCrossLanguageTransforms.class})
+  public void expandMultiOutputTest() {
+    PCollectionTuple pTuple =
+        testPipeline
+            .apply(Create.of(1, 2, 3, 4, 5, 6))
+            .apply(External.of("multi", new byte[] {}, EXPANSION_ADDR).withMultiOutputs());
+
+    PAssert.that(pTuple.get(new TupleTag<Integer>("aTag") {})).containsInAnyOrder(2, 4, 6);
+    PAssert.that(pTuple.get(new TupleTag<Integer>("bTag") {})).containsInAnyOrder(1, 3, 5);
+    testPipeline.run();
+  }
+
+  private Process runCommandline(String command) {
+    ProcessBuilder builder = new ProcessBuilder("sh", "-c", command);
+    try {
+      return builder.start();
+    } catch (IOException e) {
+      throw new AssertionError("process launch failed.");
+    }
+  }
+
+  @Test
+  @Category({ValidatesRunner.class, UsesCrossLanguageTransforms.class})
+  public void expandPythonTest() {
+    String target = "localhost:8095";
```

Comment:
Fixed ports are likely to cause problems with parallel test execution on Jenkins.